src/utilities/extendSchema.js

```diff
@@ -150,8 +152,46 @@ export function extendSchema(
         }
         directiveDefinitions.push(def);
         break;
-      case Kind.SCALAR_TYPE_EXTENSION:
       case Kind.INTERFACE_TYPE_EXTENSION:
+        const extendedInterfaceTypeName = def.name.value;
+        const existingInterfaceType = schema.getType(extendedInterfaceTypeName);
+        if (!existingInterfaceType) {
+          throw new GraphQLError(
+            `Cannot extend interface "${extendedInterfaceTypeName}" because ` +
+              'it does not exist in the existing schema.',
+            [def],
+          );
+        }
+        if (!isInterfaceType(existingInterfaceType)) {
+          throw new GraphQLError(
+            `Cannot extend non-interface type "${extendedInterfaceTypeName}".`,
+            [def],
+          );
+        }
+        // Extend all of the object types that implement this interface
```

Comment:
@lillyfwang You shouldn't do that if you want to extend interface you should also extend all objects that implement this interface with missing fields:
```graphql
extend interface SomeInterface {
  newInterface: NewInterface
}

extend type Foo {
  newInterface: NewObject
}

extend type Bar {
  newInterface: NewOtherObject
}
```
Note: similar to how you normally define a type that implements an interface you can use compatible type.  That's why you can't auto-implement interfaces 😞

Reply:
@IvanGoncharov I'm not sure if I'm understanding your comment properly, but is your concern here that "NewOtherObject" could possibly not implement "NewInterface"? Is a solution here to validate the schema after extending it to make sure the new extended schema is valid?

Reply:
@lillyfwang No I mean you can't do:
> Extend all of the object types that implement this interface

You shouldn't add fields to `GraphQLObjectType` since `extend interface` should affect only interface not object types implementing this interface.

For example, if you use bellow schema:
```graphql
interface SomeInterface {
  foo: String
}

type SomeObject {
  foo: String
}
```
Incorrect extension (correct in your PR):
```graphql
extend interface SomeInterface {
  bar: String
}
```

Correct extension:
```graphql
extend interface SomeInterface {
  bar: String
}

extend type SomeObject {
  bar: String
}
```

Reply:
Oh okay I understand - thanks for explaining!

Reply:
I agree with @IvanGoncharov here - we want to preserve the use case for extending interface fields when types already have the fields defined, and we also want to ensure we can extend objects with a more specific variant of that field, example:

```graphql
extend interface SomeInterface {
  newField: SomeInterface
}

extend type SomeObject { # Assuming it already implements SomeInterface
  newField: SomeObject # This is okay now, since SomeObject is more specific than SomeInterface
}
```